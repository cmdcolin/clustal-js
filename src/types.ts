declare interface Header {
  version: string;
  info: string;
}
declare interface Alignment {
  id: string;
  seq: string;
}

declare interface Block {
  consensus: string;
  seqs: string[];
  ids: string[];
}
declare interface Results {
  header: Header;
  alns: Alignment[];
  consensus: string;
}
