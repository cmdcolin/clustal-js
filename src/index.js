import LocalFile from './localFile'

export default class ClustalParser {
  constructor({ filename }) {
    if (filename) {
      this.clustal = new LocalFile(filename)
    }
  }

  async parse() {
    const contents = (await this.clustal.readFile()).toString().split('\n')
    const arr = contents[Symbol.iterator]()
    let line = arr.next()
    if (line.done) throw new Error('empty file')
    else line = line.value

    const knownHeaders = [
      'CLUSTAL O',
      'CLUSTAL W',
      'CLUSTALW',
      'CLUSTALO',
      'CLUSTAL',
      'PROBCONS',
      'MUSCLE',
    ]
    const header = knownHeaders.find(l => line.startsWith(l))
    const rest = line.slice(header.length)
    if (header === undefined) {
      throw new Error(
        `${header} is not a known CLUSTAL header: ${knownHeaders.join(',')}`,
      )
    }

    let version
    const words = rest.split(' ')
    for (let word of words) {
      if (word[0] === '(' && word[word.length-1] === ')') {
        word = word.substring(1, word.length - 1)
        console.log(word)
      }
      if ('0123456789'.includes(word[0])) {
        version = word
        break
      }
    }
    console.log(header, version)
    //There should be two blank lines after the header line
    line = arr.next().value
    while(line.trim() === "") {
        line = arr.next().value
    }

    const ids = []
    const seqs = []
    const consensus = ""
    let seq_cols = null

    //Use the first block to get the sequence identifiers
    while(true) {
        if(line[0] !== " " && line.trim() !== "") {
            //Sequences identifier...
            const fields = line.trimEnd().split(' ')
            console.log(fields)

            //We expect there to be two fields, there can be an optional
            //"sequence number" field containing the letter count.
            if(fields.length !== 2)
                throw new Error("Could not parse line:\n"+line)

            ids.append(fields[0])
            seqs.append(fields[1])

            //Record the sequence position to get the consensus
            if(seq_cols === null) {
                start = fields[0].length + line.slice(fields[0].length).find(fields[1])
                end = start + fields[1].length
                seq_cols = [start, end]
            }
            console.log(fields[1], line.slice(start, end))
        }//remove
    }//remove

//             if len(fields) == 3:
//                 #This MAY be an old style file with a letter count...
//                 try:
//                     letters = int(fields[2])
//                 except ValueError:
//                     raise ValueError("Could not parse line, bad sequence number:\n%s" % line)
//                 if len(fields[1].replace("-","")) != letters:
//                     raise ValueError("Could not parse line, invalid sequence number:\n%s" % line)
//         elif line[0] == " ":
//             #Sequence consensus line...
//             assert len(ids) == len(seqs)
//             assert len(ids) > 0
//             assert seq_cols is not None
//             consensus = line[seq_cols]
//             assert not line[:seq_cols.start].strip()
//             assert not line[seq_cols.stop:].strip()
//             #Check for blank line (or end of file)
//             line = handle.readline()
//             assert line.strip() == ""
//             break
//         else:
//             #No consensus
//             break
//         line = handle.readline()
//         if not line : break #end of file

    // assert line.strip() == ""
    // assert seq_cols is not None

    // #Confirm all same length
    // for s in seqs:
    //     assert len(s) == len(seqs[0])
    // if consensus:
    //     assert len(consensus) == len(seqs[0])

    // #Loop over any remaining blocks...
    // done = False
    // while not done:
    //     #There should be a blank line between each block.
    //     #Also want to ignore any consensus line from the
    //     #previous block.
    //     while (not line) or line.strip() == "":
    //         line = handle.readline()
    //         if not line : break # end of file
    //     if not line : break # end of file

    //     if line.split(None,1)[0] in known_headers:
    //         #Found concatenated alignment.
    //         done = True
    //         self._header = line
    //         break

    //     for i in range(len(ids)):
    //         assert line[0] != " ", "Unexpected line:\n%s" % repr(line)
    //         fields = line.rstrip().split()

    //         #We expect there to be two fields, there can be an optional
    //         #"sequence number" field containing the letter count.
    //         if len(fields) < 2 or len(fields) > 3:
    //             raise ValueError("Could not parse line:\n%s" % repr(line))

    //         if fields[0] != ids[i]:
    //             raise ValueError("Identifiers out of order? Got '%s' but expected '%s'" \
    //                               % (fields[0], ids[i]))

    //         if fields[1] != line[seq_cols]:
    //             start = len(fields[0]) + line[len(fields[0]):].find(fields[1])
    //             assert start == seq_cols.start, 'Old location %s -> %i:XX' % (seq_cols, start)
    //             end = start + len(fields[1])
    //             seq_cols = slice(start, end)
    //             del start, end

    //         #Append the sequence
    //         seqs[i] += fields[1]
    //         assert len(seqs[i]) == len(seqs[0])

    //         if len(fields) == 3:
    //             #This MAY be an old style file with a letter count...
    //             try:
    //                 letters = int(fields[2])
    //             except ValueError:
    //                 raise ValueError("Could not parse line, bad sequence number:\n%s" % line)
    //             if len(seqs[i].replace("-","")) != letters:
    //                 raise ValueError("Could not parse line, invalid sequence number:\n%s" % line)

    //         #Read in the next line
    //         line = handle.readline()
    //     #There should now be a consensus line
    //     if consensus:
    //         assert line[0] == " "
    //         assert seq_cols is not None
    //         consensus += line[seq_cols]
    //         assert len(consensus) == len(seqs[0])
    //         assert not line[:seq_cols.start].strip()
    //         assert not line[seq_cols.stop:].strip()
    //         #Read in the next line
    //         line = handle.readline()

    // assert len(ids) == len(seqs)
    // if len(seqs) == 0 or len(seqs[0]) == 0:
    //     raise StopIteration

    // if self.records_per_alignment is not None \
    // and self.records_per_alignment != len(ids):
    //     raise ValueError("Found %i records in this alignment, told to expect %i" \
    //                      % (len(ids), self.records_per_alignment))

    // records = (SeqRecord(Seq(s, self.alphabet), id=i, description=i) \
    //            for (i,s) in zip(ids, seqs))
    // alignment = MultipleSeqAlignment(records, self.alphabet)
    // #TODO - Handle alignment annotation better, for now
    // #mimic the old parser in Bio.Clustalw
    // if version:
    //     alignment._version = version
    // if consensus:
    //     alignment_length = len(seqs[0])
    //     assert len(consensus) == alignment_length, \
    //            "Alignment length is %i, consensus length is %i, '%s'" \
  }
}
