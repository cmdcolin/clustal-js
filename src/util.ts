export function parseVersion(line: string): string {
  const res = line.match(/\(?(\d+(\.\d+)+)\)?/);
  return res && res.length > 1 ? res[1] : "";
}

export function getFirstNonEmptyLine(arr: Symbol.iterator): Results {
  // There should be two blank lines after the header line
  let line = arr.next().value;
  while (line !== undefined && line.trim() === "") {
    line = arr.next().value;
  }
  return line;
}
