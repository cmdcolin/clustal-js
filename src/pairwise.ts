import { getFirstNonEmptyLine } from './util'

export function getSeqBounds(line: string) {
  const fields = line.split(/\s+/)
  const k1 = fields[0].length + fields[1].length
  const temp = line.slice(k1)
  const s = k1 + temp.indexOf(fields[2])
  const e = s + fields[2].length
  return [s, e] as const
}

// Use the first block to get the sequence identifiers
export function parsePairwiseBlock(arr: Iterator<string>) {
  let line = getFirstNonEmptyLine(arr)
  const block = []
  let consensusLine = ''
  if (!line) {
    return undefined
  }

  while (line) {
    if (line.startsWith(' ')) {
      consensusLine = line
    } else {
      block.push(line)
    }
    line = arr.next().value
  }
  const [start, end] = getSeqBounds(block[0])
  const fields = block.map(s => s.split(/\s+/))
  const ids = fields.map(s => s[0])
  const seqs = fields.map(s => s[2])
  let consensus = consensusLine.slice(start, end)

  // handle if the consensus trailing whitespace got trimmed
  const remainder = seqs[0].length - consensus.length
  if (remainder) {
    consensus += ' '.repeat(remainder)
  }
  return { ids, seqs, consensus }
}

export function parsePairwiseBlocks(arr: Iterator<string>) {
  let block
  const res = parsePairwiseBlock(arr)
  if (res !== undefined) {
    while ((block = parsePairwiseBlock(arr))) {
      for (let i = 0; i < block.seqs.length; i++) {
        res.seqs[i] += block.seqs[i]
      }
      res.consensus += block.consensus
    }
  }
  return res
}
