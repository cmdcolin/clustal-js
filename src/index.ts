import { parsePairwiseBlocks } from './pairwise'
import { parseBlocks, parseHeader, getFirstNonEmptyLine } from './util'

export function parseClustalIter(arr: Iterator<string>) {
  const line = getFirstNonEmptyLine(arr)
  if (!line) {
    throw new Error('Empty file received')
  }
  const header = parseHeader(line)

  const res = parseBlocks(arr)
  if (res === undefined) {
    throw new Error('No blocks parsed')
  }

  const alns = res.seqs.map((n, index) => ({ id: res.ids[index], seq: n }))
  const { consensus } = res
  if (consensus.length != alns[0].seq.length) {
    throw new Error(
      `Consensus length != sequence length. Con ${consensus.length} seq ${alns[0].seq.length}`,
    )
  }

  return { consensus, alns, header }
}

interface Row {
  id: string
  seq: string
}
export function parsePairwiseIter(arr: string) {
  const res = parsePairwiseBlocks(arr.split('\n')[Symbol.iterator]())
  if (res === undefined) {
    throw new Error('No blocks parsed')
  }

  const alns = res.seqs.map((n, index) => ({ id: res.ids[index], seq: n }))
  const { consensus } = res
  if (consensus.length != alns[0].seq.length) {
    throw new Error(
      `Consensus length != sequence length. Con ${consensus.length} seq ${alns[0].seq.length}`,
    )
  }
  if (alns.length !== 2) {
    throw new Error('More than two sequences in pairwise alignment')
  }
  return {
    consensus,
    alns: alns as [Row, Row],
  }
}

export function parse(contents: string) {
  const iter = contents.split('\n')[Symbol.iterator]()
  return parseClustalIter(iter)
}

export function parsePairwise(contents: string) {
  return parsePairwiseIter(
    contents
      .split('\n')
      .filter(f => !f.startsWith('#'))
      .join('\n'),
  )
}
