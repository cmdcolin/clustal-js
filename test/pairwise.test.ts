import { test, expect } from 'vitest'
import { parsePairwise } from '../src'
import * as fs from 'fs'

test('pairwise', async () => {
  const ret = fs.readFileSync('test/data/pairwise.aln', 'utf8')
  const res = parsePairwise(ret)
  expect(res).toMatchSnapshot()
})
