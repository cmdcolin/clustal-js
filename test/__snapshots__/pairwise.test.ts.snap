// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pairwise 1`] = `
{
  "alns": [
    {
      "id": "a",
      "seq": "MGQKGHKDSLYPCGGTPESSLHEALDQCMTALDLFLTNQFSEALSYLKPRTKESMYHSLTYATILEMQAMMTFDPQDILLAGNMMKEAQMLCQRHRRKSSVTDSFSSLVNRPTLGQFTEEEIHAEVCYAECLLQRAALTFLQGSSHGGAVRPRALHDPSHACSCPPGPGRQHLFLLQDENMVSFIKGGIKVRNSYQTYKELDSLVQSSQYCKGENHPHFEGGVKLGVGAFNLTLSMLPTRILRLLEFVGFSGNKDYGLLQLEEGASGHSFRSVLCVMLLLCYHTFLTFVLGTGNVNIEEAEKLLKPYLNRYPKGAIFLFFAGRIEVIKGNIDAAIRRFEECCEAQQHWKQFHHMCYWELMWCFTYKGQWKMSYFYADLLSKENCWSKATYIYMKAAYLSMFGKEDHKPFGDDEVELFRAVPGLKLKIAGKSLPTEKFAIRKSRRYFSSNPISLPVPALEMMYIWNGYAVIGKQPKLTDGILEIITKAEEMLEKGPENEYSVDDECLVKLLKGLCLKYLGRVQEAEENFRSISANEKKIKYDHYLIPNALLELALLLMEQDRNEEAIKLLESAKQNYKNYSMESRTHFRIQAATLQAKSSLENSSRSMVSSVSL-",
    },
    {
      "id": "b",
      "seq": "--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------MKAAYLSMFGKEDHKPFGDDEVELFRAVPGLKLKIAGKSLPTEKFAIRKSRRYFSSNPISLPVPALEMMYIWNGYAVIGKQPKLTDGILEIITKAEEMLEKGPENEYSVDDECLVKLLKGLCLKYLGRVQEAEENFRSISANEKKIKYDHYLIPNALLELALLLMEQDRNEEAIKLLESAKQNYKNYSMESRTHFRIQAATLQAKSSLENSSRSMVSSVSL*",
    },
  ],
  "consensus": "                                                                                                                                                                                                                                                                                                                                                                                                        ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ",
}
`;
